lib/node4Migration.js

```diff
@@ -0,0 +1,161 @@
+                }
+                else if (r.body.status === 'failed') {
+                    return cb(new Error(r.body.message), r.body);
+                }
+                else {
+                    process.stdout.write('.');
+                    return setTimeout(() => tryMigrate(remainingAttempts - 1), 3000);
+                }
+            });
+    }
+};
+
+exports.deleteNode4Assets = function (options) {
+
+    var profile = new Sandbox({
+        url: 'https://webtask.it.auth0.com',
+        container: options.container,
+        token: options.token
+    });
+
+    return new Promise(function (resolve, reject) {
+        return Async.series([
+            (cb) => determineLegacyStack(cb),
+            (cb) => deleteCronJobs(cb),
+            (cb) => deleteWebtasks(cb)
+        ], (e) => e ? reject(e) : resolve(null));
+    });
+
+    function determineLegacyStack(cb) {
+        // This is to determine if the DNS has been repointed to the new stack already,
+        // in which case no deletion should occurr.
+        return Superagent.get('https://webtask.it.auth0.com/health')
+            .end((e,r) => {
+                profile.isLegacyNode4Stack = !e && r && r.body && r.body.stack_name && r.body.stack_name.indexOf('wa-usw-') === 0;
+                return cb(e);
+            });
+    }
+
+    function deleteCronJobs(cb) {
+        if (!profile.isLegacyNode4Stack) return cb();
+
+        return Async.waterfall([
+            (cb) => getAsset('listCronJobs', cb),
+            (crons, cb) => doDelete(crons, cb)
+        ], cb);
+
+        function doDelete(crons, cb) {
+            if (crons.length === 0) {
+                console.log('There are no Node 4 CRON jobs to delete.');
+            }
+            else {
+                console.log('Deleting Node 4 CRON jobs:');
+            }
+            return Async.eachSeries(crons, deleteCron, cb);
+        }
+    }
+
+    function deleteWebtasks(cb) {
+        if (!profile.isLegacyNode4Stack) return cb();
+
+        return Async.waterfall([
+            (cb) => getAsset('listWebtasks', cb),
+            (webtasks, cb) => doDelete(webtasks, cb)
+        ], cb);
+
+        function doDelete(webtasks, cb) {
+            if (webtasks.length === 0) {
+                console.log('There are no Node 4 webtasks to delete.');
+            }
+            else {
+                console.log('Deleting Node 4 webtasks:');
+            }
+            return Async.eachSeries(webtasks, deleteWebtask, cb);
+        }
+    }
+
+    function getAsset(method, cb) {
+
+        var result = [];
+        return append();
+
+        function append() {
+            profile[method]({ 
+                offset: result.length, 
+                limit: 100 
+            }, (e, d) => {
+                if (e) return cb(e);
+                if (!d || d.length === 0) return cb(null, result);
+                d.forEach(w => result.push(w.toJSON().name));
+                append();
```

Comment:
This will always do one extra round trip. I guess it is safer but wondering if it was intended.

Reply:
Same comment as in the other place. Better safe than sorry.